btree/btree.c

```diff
@@ -5358,6 +5470,215 @@ static void m0_btree_ut_mt_mt_kv_oper(void)
+#define MAX_RECS_FOR_TREE_TEST  100
+
+/**
+ * This routine is a thread handler which primarily involves in creating,
+ * opening, closing and destroying btree. To run out-of-sync with other threads
+ * it also launches PUT, GET, ITER and DEL operations on the btree for a random
+ * count.
+ */
+static void btree_ut_tree_oper_thread_handler(struct btree_ut_thread_info *ti)
+{
+	uint64_t               key;
+	uint64_t               value;
+	m0_bcount_t            ksize = sizeof key;
+	m0_bcount_t            vsize = sizeof value;
+	void                  *k_ptr = &key;
+	void                  *v_ptr = &value;
+	struct m0_btree_rec    rec   = {
+				     .r_key.k_data = M0_BUFVEC_INIT_BUF(&k_ptr,
+									&ksize),
+				     .r_val        = M0_BUFVEC_INIT_BUF(&v_ptr,
+									&vsize),
+				     .r_flags      = 0,
+				     };
+	struct cb_data         data  = {
+					.key         = &rec.r_key,
+					.value       = &rec.r_val,
+					.check_value = false,
+					.flags       = 0,
+				       };
+	struct m0_btree_cb     ut_cb   = {
+					  .c_act       = btree_kv_put_cb,
+					  .c_datum     = &data,
+					 };
+	int32_t                loop_count;
+	struct m0_btree_op     kv_op     = {};
+	void                  *temp_node;
+	struct m0_btree_type   btree_type  = {.tt_id = M0_BT_UT_KV_OPS,
+					      .ksize = sizeof(key),
+					      .vsize = sizeof(value),
+					     };
+	const struct node_type *nt         = &fixed_format;
+	struct m0_be_tx        *tx         = NULL;
+
+	random_r(&ti->ti_random_buf, &loop_count);
+	loop_count %= (MAX_TREE_LOOPS - MIN_TREE_LOOPS);
+	loop_count += MIN_TREE_LOOPS;
+
+	/** Create temp node space and use it as root node for btree */
+	temp_node = m0_alloc_aligned((1024 + sizeof(struct nd)), 10);
+
+	while (loop_count--) {
+		struct m0_btree_op  b_op        = {};
+		struct m0_btree    *tree;
+		int32_t             rec_count;
+		uint32_t            i;
+
+		/**
+		 * 1) Create a tree
+		 * 2) Add a few random count of records in the tree.
+		 * 3) Close the tree
+		 * 4) Open the tree
+		 * 5) Confirm the records are present in the tree.
+		 * 6) Close the tree
+		 * 4) Open the tree
+		 * 5) Delete all the records from the tree.
+		 * 6) Close the tree
+		 * 7) Destroy the tree
+		 */
+
+		M0_BTREE_OP_SYNC_WITH_RC(&b_op.bo_op,
+					 m0_btree_create(temp_node, 1024,
+							 &btree_type, nt, tx,
+							 &b_op),
+					 &b_op.bo_sm_group, &b_op.bo_op_exec);
+
+		tree = b_op.bo_arbor;
+
+		random_r(&ti->ti_random_buf, &rec_count);
+		rec_count %= MAX_RECS_FOR_TREE_TEST;
+		rec_count = rec_count ? : (MAX_RECS_FOR_TREE_TEST / 2);
+
+		for (i = 1; i <= rec_count; i++) {
+			key   = i;
+			value = key;
+
+			M0_BTREE_OP_SYNC_WITH_RC(&kv_op.bo_op,
+						 m0_btree_put(tree, tx,
+							      &rec, &ut_cb,
+							      0, &kv_op),
+						 &kv_op.bo_sm_group,
+						 &kv_op.bo_op_exec);
+			M0_ASSERT(data.flags == M0_BSC_SUCCESS);
+		}
+
+		m0_btree_close(tree);
+
+		m0_btree_open(temp_node, 1024, &tree);
```

Comment:
We can add return code check (ASSERT) for tree/open close.

Reply:
I will add for btree_open but cannot add for btree_close since it returns a void.